src/rust/src/sapling.rs

```diff
@@ -0,0 +1,176 @@
+// Copyright (c) 2020-2022 The Zcash developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or https://www.opensource.org/licenses/mit-license.php .
+
+// This is added because `check_spend` takes several arguments over FFI. This
+// annotation gets removed by the cxx procedural macro so it needs to be enabled
+// on the entire module.
+#![allow(clippy::too_many_arguments)]
+
+use bellman::groth16::Proof;
+use group::GroupEncoding;
+
+use zcash_primitives::{
+    sapling::redjubjub::{self, Signature},
+    transaction::components::Amount,
+};
+use zcash_proofs::sapling::SaplingVerificationContext;
+
+use super::GROTH_PROOF_SIZE;
+use super::{de_ct, SAPLING_OUTPUT_VK, SAPLING_SPEND_VK};
+
+#[cxx::bridge(namespace = "sapling")]
+mod ffi {
+    extern "Rust" {
+        type Verifier;
+
+        fn init_verifier() -> Box<Verifier>;
+        fn check_spend(
+            &mut self,
+            cv: &[u8; 32],
+            anchor: &[u8; 32],
+            nullifier: &[u8; 32],
+            rk: &[u8; 32],
+            zkproof: &[u8; 192], // GROTH_PROOF_SIZE
```

Comment:
[Not blocking] Can we use the actual `GROTH_PROOF_SIZE` constant here, or Rust doesn't allow that?

Reply:
It appears that even though Rust would ordinarily be fine with it, the cxx::bridge macro doesn't support const generics yet.